### ISSUE TITLE
Hashing/Encoding Resource Identity 

### ISSUE THREAD

**Issue by circleupx** [scott-mcdonald/JsonApiFramework#78]
Scott, 

Does JsonApiFramework currently has the ability to override the value of an id property on a resource object?

For example, take the following resource. 

```json
HTTP/1.1 200 OK
Content-Type: application/vnd.api+json

{
  "data": {
    "type": "articles",
    "id": "1",
    "attributes": {
      "title": "JSON:API paints my bikeshed!"
    },
    "relationships": {
      "author": {
        "data": {"id": "42", "type": "people"}
      }
    }
  }
}
```

I would like to encode/hash the value of all id properties, the resource above may now look like this.

```json
HTTP/1.1 200 OK
Content-Type: application/vnd.api+json

{
  "data": {
    "type": "articles",
    "id": "U6dc",
    "attributes": {
      "title": "JSON:API paints my bikeshed!"
    },
    "relationships": {
      "author": {
        "data": {"id": "u87U", "type": "people"}
      }
    }
  }
}
```

Basically, I want to prevent users from being able to guess/see internal ids without having to change the database identifiers.

**Comment by scott-mcdonald** [comment]
@circleupx I can think of 2 approaches to doing this:
- Having an an alternative computed property that is the hash of the primary key and configure the alternative computer property as the resource identity of the resource for JsonApiFramework and ignore the sequential primary key property with respect to JsonApiFramework.
- Having 2 properties where one is the primary key the other is the hash and you generate the hash once when the entity is created itself. Then make the hash the resource identity of the resource for JsonApiFramework and ignore the sequential primary key property with respect to JsonApiFramework.

Basically similar approaches where one the alternative primary key is a computed property, the other you make it a full blown property that is set once when the resource is created.

I might have a need for this in the future myself so let me know how it goes. The other approach would be literally add it as a feature of JsonApiFramework similar to how EntityFramework allows a primary key and alternate primary key...

**Comment by circleupx** [comment]
I will try both suggestions to which works best for my use case.

> The other approach would be literally add it as a feature of JsonApiFramework similar to how EntityFramework allows a primary key and alternate primary key...

I was thinking that it could be a feature of JsonApiFramework. I know you are busy with work if you like I could take a shot at adding this feature, I would just need to know how you would like it to be implemented or if you prefer that I come up with the design. 

Let me know. 

Thanks!

**Comment by scott-mcdonald** [comment]
Well to be honest this definitely would be an involved feature so on the larger/complex scope for sure. I would approach it along the lines of how EntityFrameworkCore does it. At the metadata level I would define an "alternative key"; the question then comes in when to use the alternative key versus the main key for the "id" resource identity property. Need to think about this feature for sure as the framework definitely uses identity throughout the document either building or reading process...

**Comment by circleupx** [comment]
@scott-mcdonald 

Hey, I ended up going with your first approach. 

- Having an alternative computed property that is the hash of the primary key and configures the alternative computer property as the resource identifier of the resource for JsonApiFramework and ignore the sequential primary key property with respect to JsonApiFramework.

I ended up using https://github.com/ullmark/hashids.net. It all works, the only drawback was that I needed to handle the hash ids on post/patch requests. You can unhash the id on each post/patch method or build a middleware to handle the transformation. I believe you said you are already doing something like that for query parameters.
